pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/internal/Helper.java

```diff
@@ -0,0 +1,230 @@
+                        return Helper.isMethodName(methodNode, newClassName[0], methodName);
+                    }
+                }
+            }
+        }
+
+        return false;
+    }
+
+    public static String getFQVariableName(final ASTVariableExpression variable) {
+        final ASTReferenceExpression ref = variable.getFirstChildOfType(ASTReferenceExpression.class);
+        String objectName = "";
+        if (ref != null) {
+            if (ref.getNode().getNames().size() == 1) {
+                objectName = ref.getNode().getNames().get(0).getValue() + ".";
+            }
+        }
+
+        VariableExpression n = variable.getNode();
+        StringBuilder sb = new StringBuilder().append(n.getDefiningType().getApexName()).append(":").append(objectName)
+                .append(n.getIdentifier().getValue());
+        return sb.toString();
+    }
+
+    public static String getFQVariableName(final ASTVariableDeclaration variable) {
+        VariableDeclaration n = variable.getNode();
+        StringBuilder sb = new StringBuilder().append(n.getDefiningType().getApexName()).append(":")
+                .append(n.getLocalInfo().getName());
+        return sb.toString();
+    }
+
+    public static String getFQVariableName(final ASTField variable) {
+        Field n = variable.getNode();
+        StringBuilder sb = new StringBuilder()
+                .append(n.getDefiningType().getApexName()).append(":")
+                .append(n.getFieldInfo().getName());
+        return sb.toString();
+    }
+
+    static String getVariableType(final ASTField variable) {
+        Field n = variable.getNode();
+        StringBuilder sb = new StringBuilder().append(n.getDefiningType().getApexName()).append(":")
+                .append(n.getFieldInfo().getName());
+        return sb.toString();
+    }
+    
+    public static String getFQVariableName(final ASTFieldDeclaration variable) {
+        StringBuilder sb = new StringBuilder()
+                .append(variable.getNode().getDefiningType().getApexName()).append(":")
+                .append(variable.getImage());
+        return sb.toString();
+    }
+
+    public static String getFQVariableName(final ASTNewKeyValueObjectExpression variable) {
+        StringBuilder sb = new StringBuilder()
+                .append(variable.getNode().getDefiningType().getApexName()).append(":")
+                .append(variable.getType());
+        return sb.toString();
+    }
+
+    public static boolean isSystemLevelClass(ASTUserClass node) {
+        List<TypeRef> interfaces = node.getNode().getDefiningType().getCodeUnitDetails().getInterfaceTypeRefs();
+
+        for (TypeRef intObject : interfaces) {
+            if (isWhitelisted(intObject.getNames())) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private static boolean isWhitelisted(List<Identifier> ids) {
+        StringBuffer sb = new StringBuffer();
+
+        for (int i = 0; i < ids.size(); i++) {
```

Comment:
Isn't the a more redable for each in Java?

Reply:
Not if you need access to the index. What's done here is basically joining the identifiers with ".". With java8 (and I think, we are here in Apex already with java8) we can use Streams to do that.